Add support for geofence and rally point upload

ken-voly opened the pull request:
This pull request adds geofence and rally points to the API created in #1343.  The following API functions were added:
- UploadGeofence()
- UploadRallyPoints()

These changes are intended to be used with the proto changes here https://github.com/volansi/MAVSDK-Proto/tree/pr-add-rally-geofence

## TODO
- Clean up new functions
- Improve error handling

JonasVautherin reviewed (commented):
Thanks! Two questions:

1. Is your `UploadGeofence` different from the one of the [`geofence` plugin](https://github.com/mavlink/MAVSDK-Proto/blob/main/protos/geofence/geofence.proto#L16)?
2. Why not having those changes directly inside the `geofence` plugin?

ken-voly commented:
>     1. Is your `UploadGeofence` different from the one of the [`geofence` plugin](https://github.com/mavlink/MAVSDK-Proto/blob/main/protos/geofence/geofence.proto#L16)?

The `UploadGeofence` in `mission_raw` takes a list of mission items as a parameter. These mission items are returned by `ImportQgroundcontrolMission` in `mission_raw`.

The `UploadGeofence` in `geofence` takes a list of `Polygon` an in input parameter. Also, it looks like this method may not support circular geofences.

>     2. Why not having those changes directly inside the `geofence` plugin?

It's up for debate on where these changes should be added. From previous discussion, it sounded like mission uploading (from a file) should go into `mission_raw`. Because the geofence is from a mission plan, I put it in `mission_raw`.

JonasVautherin commented:
I see. Then I guess it makes sense there, yes :+1:

julianoes commented:
Sorry, this stalled! Should we try to get it in soon?

ken-voly commented:
> Sorry, this stalled! Should we try to get it in soon?

No problem. I actually wanted to improve the error handling before having it merged, but I haven't had much time. I'll see if I can get around to it.

Also, I think the issues fixed with 6e2afe01c97c05270d8ba1127d183b4f4e8dce88 may have been due to an incorrect setup on a specific machine. Can you confirm that you can compile MAVSDK without that commit?

julianoes commented:
Commit https://github.com/mavlink/MAVSDK/commit/6e2afe01c97c05270d8ba1127d183b4f4e8dce88 seems to be related to a mavlink change, I think. Let me try to update mavlink and see if I need to cherry-pick it.

julianoes commented:
I did an update to mavlink here but I did not see these issues: https://github.com/mavlink/MAVSDK/pull/1670

julianoes commented on src/plugins/mission_raw/mission_import.cpp:
```diff
@@ -110,6 +123,122 @@ MissionImport::import_mission(const Json::Value& root)
     return {mission_items};
 }
 
+static float set_float(const Json::Value& val)
+{
+    return val.isNull() ? NAN : val.asFloat();
+}
+
+static float set_int32(const Json::Value& val)
+{
+    return static_cast<int32_t>(val.isNull() ? 0 : (std::round(val.asDouble() * 1e7)));
+}
```
I would have these as `static` private class member functions instead, just for taste.

julianoes reviewed (commented):
Small comment, other than that, this looks fine. Make sure to fix the conflicts too, not sure what GitHub is complaining there.

ken-voly commented:
@julianoes I rebased on the current `main` branch and did a little cleanup. 

I don't have the time or knowledge to do proper input checking of the mission plan right now. From what I could tell, some fields are optional and some aren't. For example, I found that certain versions of QGC may not include the "inclusion" field when creating a geofence. For now, the rally point and geofence fields are not thoroughly checked.

The below commit adds the needed functions to the proto repository. I don't know where to merge the proto changes so this PR no longer changes the version of the proto submodule. We can add it back in later if needed.
https://github.com/volansi/MAVSDK-Proto/tree/63e58aefdf2659651e29d02d38a49ae36cd4f311

julianoes commented:
I will review this soon, thank you!

julianoes commented on src/mavsdk/plugins/mission_raw/mission_import.cpp:
```diff
@@ -110,6 +123,102 @@ MissionImport::import_mission(const Json::Value& root)
+    // Return early if there are no geofence items.
+    const auto geofence = root["geoFence"];
+    if (geofence.empty()) {
+        return {geofence_items};
```
I think here we want to return `std::nullop`, right?

julianoes commented on src/mavsdk/plugins/mission_raw/mission_import.cpp:
```diff
@@ -110,6 +123,102 @@ MissionImport::import_mission(const Json::Value& root)
+    // Return early if there are no rally points.
+    const auto rally_points = root["rallyPoints"];
+    if (rally_points.empty()) {
+        return {rally_items};
```
Same question here... :thinking:

julianoes commented on src/mavsdk/plugins/mission_raw/mission_import.cpp:
```diff
@@ -219,4 +320,66 @@ MissionImport::import_complex_mission_items(const Json::Value& json_item)
+    std::vector<MissionRaw::MissionItem> circular_geofences;
+
+    for (const auto& circle : circles) {
+        bool inclusion = circle["inclusion"].asBool();
```
What should we do if there is no field `inclusion`? I think we should default to `inclusion = true` instead, no?

ken-voly replied:
Okay. If the field is missing, I'll make it default to true.

julianoes commented:
@ken-voly I don't have permission to push to the branch of this PR.

Do you mind pointing the proto submodule to https://github.com/mavlink/MAVSDK-Proto/pull/282 and then rebasing and force-pushing again?

julianoes reviewed (commented):
Looks good, just 3 small comments.

julianoes commented:
@ken-voly it would be great if you could check my questions on this PR. Also, I'm not allowed to push to this PR to fix it up, so I will soon just cherry-pick your commits into another PR and finish it off.

ken-voly commented:
@julianoes I'll review your comments today.

julianoes commented:
As soon as you have fixed this PR and CI is looking good, then we can merge https://github.com/mavlink/MAVSDK-Proto/pull/282, point the submodule to main again, and merge this.

In the next push, I suggest you point the submodule to https://github.com/mavlink/MAVSDK-Proto/pull/282.

ken-voly commented:
I made the recommended fixes, updated the submodule version, and rebased on `main`. Once the submodule PR is merged, I'll update the submodule version again and we can merge this PR.

julianoes commented:
@ken-voly it looks like some import tests are failing in CI. If you could have a look next week, that would be great.

julianoes commented:
Replaced by #1892. Thanks @ken-voly, sorry for the late reviews and fixing it up but it will eventually make it in!
